### PR TITLE
[TG-5572] Add option to restrict numerical input values

### PR DESCRIPTION
**Pull request by majakusber**
Adds command line option `--java-assume-input-limits [int:int]`. Does not work for arrays yet.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (romainbrenguier) [inline jbmc/regression/jbmc/assume-input-limits/My.java]: this line is too long

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: :pick: I would prefer `type != java_bool_type() && type != java_char_type()`

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: :pick: maybe do `if(const auto &lower = ...)` so that you can use `lower.value()` below instead of the long expression

File: src/util/object_factory_parameters.cpp
Comment (romainbrenguier) [inline src/util/object_factory_parameters.cpp]: typo "pf" -> of

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 01f3e63a3673746ccb08b09c0cec50d3ecba47db).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115054010

File: src/util/object_factory_parameters.cpp
Comment (romainbrenguier) [inline src/util/object_factory_parameters.cpp]: I think using a regex `"\\[(\\d*):(\\d*)\\]"` would be a bit clearer and avoid having to think about indexes

File: src/util/object_factory_parameters.h
Comment (romainbrenguier) [inline src/util/object_factory_parameters.h]: I'm not sure these two need to be grouped in a struct type, since the type itself doesn't seem to be used outside of this field.

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (thk123) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: This behaviour should be documented in the command line options help

File: src/util/object_factory_parameters.cpp
Comment (thk123) [inline src/util/object_factory_parameters.cpp]: :no_entry_sign: User errors should throw an exception rather than an invariant. Specifically `user_input_error_exceptiont`

File: src/util/object_factory_parameters.h
Comment (thk123) [reply src/util/object_factory_parameters.h]: I disagree - they are logically one thing (a range). However, does it make sense to have a lower but not an upper or vice versa? I'd probably make the whole thing optional if you only support having both or neither set

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (thk123) [inline jbmc/regression/jbmc/assume-input-limits/My.java]: :pick: put the ifs in the same order as the parameters

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (thk123) [inline jbmc/regression/jbmc/assume-input-limits/My.java]: :pick: Ideally this would go in the other (Array) commit to ease reviewing

File: jbmc/regression/jbmc/assume-input-limits/arg_assume_no_zero.desc
Comment (thk123) [inline jbmc/regression/jbmc/assume-input-limits/arg_assume_no_zero.desc]: :question: Not clear on what this test checks? There is no assert that can be reached by only using 0?

**Review by thk123 — COMMENTED** [review]
:no_entry_sign: Please add some tests for invalid format of parameter and check the exit code is `1` :white_check_mark:

File: src/util/object_factory_parameters.h
Comment (majakusber) [reply src/util/object_factory_parameters.h]: I also find it more understandable if they are grouped. 

Originally, I had the whole thing as optional but I think supporting only one limit could be quite useful (allows to for example force using only positive values). Alternatively, at one point I had default values for the limits - min and max long. However, asking whether the optional has a value is I think nicer than comparing the value to max long. I also thought of making the whole thing as well as the individual limits optional but that seems like an overkill :slightly_smiling_face:

File: src/util/object_factory_parameters.h
Comment (tautschnig) [inline src/util/object_factory_parameters.h]: Is `int` really the right type? I'd use `mp_integer` here as the type of the system JBMC is being compiled on is unrelated to the type that you want to talk about.

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (majakusber) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: Done

File: src/util/object_factory_parameters.cpp
Comment (majakusber) [reply src/util/object_factory_parameters.cpp]: Corrected

File: src/util/object_factory_parameters.cpp
Comment (majakusber) [reply src/util/object_factory_parameters.cpp]: Done, I used `invalid_command_line_argument_exceptiont`

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (majakusber) [reply jbmc/src/java_bytecode/java_object_factory.cpp]: Done and done

File: src/util/object_factory_parameters.cpp
Comment (majakusber) [reply src/util/object_factory_parameters.cpp]: Done

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (majakusber) [reply jbmc/regression/jbmc/assume-input-limits/My.java]: Done

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (majakusber) [reply jbmc/regression/jbmc/assume-input-limits/My.java]: Shortened

File: jbmc/regression/jbmc/assume-input-limits/My.java
Comment (majakusber) [reply jbmc/regression/jbmc/assume-input-limits/My.java]: Done

File: jbmc/regression/jbmc/assume-input-limits/arg_assume_no_zero.desc
Comment (majakusber) [reply jbmc/regression/jbmc/assume-input-limits/arg_assume_no_zero.desc]: I changed the tests a bit - all desc files check for positive limits only `[1:3]` and there is a specific test fro checking that negative limit works too `arg_assume_neg`

File: src/util/object_factory_parameters.h
Comment (majakusber) [reply src/util/object_factory_parameters.h]: Good point, done!

**Comment by majakusber** [comment]
@romainbrenguier @thk123 @tautschnig All comments addressed, please have another look. I left most of the changes in the `Fixups`/`Fixup` commits but will squash those before merging.

General question regarding the option name: would you prefer `--java-assume-input-limits` or `--java-assume-input-bounds`?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 56a84cfd4a5cd859723af9167bf1e36eb0ddc40c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115267886

File: src/util/object_factory_parameters.cpp
Comment (romainbrenguier) [inline src/util/object_factory_parameters.cpp:56]: :+1:

File: src/util/object_factory_parameters.h
Comment (romainbrenguier) [inline src/util/object_factory_parameters.h]: The documentation should also tell what the type represents not how it will be used.

File: src/util/object_factory_parameters.h
Comment (romainbrenguier) [reply src/util/object_factory_parameters.h]: In that case "range" or "interval" should probably appear in the name. 
Rather that defining a new type you should reuse existing ones, here `integer_intervalt` would be the right one. It will avoid having to define a new type and you can make your code simpler by using `.empty()`.

**Review by romainbrenguier — APPROVED** [review]

File: src/util/object_factory_parameters.h
Comment (majakusber) [reply src/util/object_factory_parameters.h]: Oh yes! That looks like exactly what I need, thank you :+1:

**Comment by majakusber** [comment]
@tautschnig @thk123 Ready for re-review, as @romainbrenguier suggested I used `integer_intervalt` instead of my own struct, I renamed the option to match this and also to match the related `java-assume-inputs-non-null` option, and I added tests for static fields.

File: src/util/object_factory_parameters.h
Comment (romainbrenguier) [inline src/util/object_factory_parameters.h]: I think we shouldn't have dependencies of util on analyses, so the best solution is probably to move `interval_template.h` to  `util` (and we can drop "template" from the name)

File: jbmc/src/java_bytecode/java_bytecode_language.h
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_language.h:62]: :bulb: I'd phrase this as the positive "only initialize primitive typed parameters to values within the specified range"

**Review by thk123 — APPROVED** [review]
lgtm

File: src/util/object_factory_parameters.h
Comment (majakusber) [reply src/util/object_factory_parameters.h]: Yep, moved... with a bunch of other files it needed...

**Review by tautschnig — CHANGES_REQUESTED** [review]
I'm afraid the move of files is a bit too far-reaching. It seems that all that you need is `integer_intervalt`, so I guess that should just be placed in a file of its own. The `ai*.h` files really shouldn't be moved.

File: unit/analyses/ai/ai.cpp
Comment (romainbrenguier) [inline unit/analyses/ai/ai.cpp]: the header file and cpp file shouldn't be in separate modules and I think ai should stay in analyses

File: jbmc/src/java_bytecode/java_bytecode_language.h
Comment (majakusber) [reply jbmc/src/java_bytecode/java_bytecode_language.h:62]: Rephrased

**Comment by majakusber** [comment]
I minimized the move :crossed_fingers:

File: src/util/integer_interval.h
Comment (romainbrenguier) [inline src/util/integer_interval.h]: :pick: `using` is preferred

**Review by romainbrenguier — APPROVED** [review]

File: src/util/integer_interval.h
Comment (majakusber) [reply src/util/integer_interval.h]: Done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f303e99ca7afca324b27498bc6548902db99d4fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115425794

**Comment by majakusber** [comment]
@tautschnig Would you have time to give a full review? I need a code-owner :)

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_object_factory.cpp:1016]: Why is `java_char_type()` also special cased? It would seem that intervals still make sense for that type?

File: src/util/object_factory_parameters.cpp
Comment (tautschnig) [inline src/util/object_factory_parameters.cpp:66]: Why not `!base_match[1].empty()`

File: src/util/object_factory_parameters.cpp
Comment (tautschnig) [inline src/util/object_factory_parameters.cpp:68]: Why not `!base_match[2].empty()`

**Review by tautschnig — APPROVED** [review]

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_object_factory.cpp:1016]: it makes sense but we would probably want a different one, for instance if we use [0:10] for integers it would only give non-printable characters

Comment (majakusber) [reply jbmc/src/java_bytecode/java_object_factory.cpp:1016]: I think that wouldn't be very intuitive for the user (definitely not for me :slightly_smiling_face: ). We have `--string-input-value` option which accepts a list of strings to use during test generation (it's not exactly the same though as what you meant).

File: src/util/object_factory_parameters.cpp
Comment (majakusber) [reply src/util/object_factory_parameters.cpp:66]: The type doesn't have `empty()` function. Note that because the regex allows empty string here, this would always be a match. I just need to check that what matched was an empty string (no bound).

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (thk123) [reply jbmc/src/java_bytecode/java_object_factory.cpp:1016]: A little forward visibility - a forthcoming feature will bound `java_char_type` when `--string-printable` is used (using the same range of characters).

**Pull request merged**
